File: frontend/src/services/knowledge.ts
import { requestJson } from "@/services/http"
import type { ProductManual, ProductSpec, FaultSOP } from "@/types/knowledge"

export interface UploadManualPayload {
  product_id: string
  title: string
}

export interface ImportSpecsPayload {
  product_id: string
  source: string
}

export interface CreateSOPPayload {
  product_id: string
  fault_type: string
  fault_category: string
  symptoms: Array<{ keyword: string; weight: number }>
  steps: Array<{ order: number; action: string; expected: string; next_if_pass: number; next_if_fail: number }>
  resolution: string
}

export async function fetchManuals(productId: string): Promise<ProductManual[]> {
  const response = await requestJson<ProductManual[]>(`/api/knowledge/manuals?product_id=${productId}`)

  if (!response.data) {
    throw new Error("服务未返回手册列表。")
  }

  return response.data
}

export async function uploadManual(
  file: File,
  payload: UploadManualPayload,
): Promise<ProductManual> {
  const formData = new FormData()
  formData.append("file", file)
  formData.append("product_id", payload.product_id)
  formData.append("title", payload.title)

  const response = await requestJson<ProductManual>("/api/knowledge/manuals/upload", {
    method: "POST",
    body: formData,
  })

  if (!response.data) {
    throw new Error("上传手册成功，但服务未返回手册信息。")
  }

  return response.data
}

export async function fetchSpecs(productId: string): Promise<ProductSpec[]> {
  const response = await requestJson<ProductSpec[]>(`/api/knowledge/specs?product_id=${productId}`)

  if (!response.data) {
    throw new Error("服务未返回规格列表。")
  }

  return response.data
}

export async function importSpecs(payload: ImportSpecsPayload): Promise<ProductSpec> {
  const response = await requestJson<ProductSpec>("/api/knowledge/specs/import", {
    method: "POST",
    body: JSON.stringify(payload),
  })

  if (!response.data) {
    throw new Error("导入规格成功，但服务未返回规格信息。")
  }

  return response.data
}

export async function fetchSOPs(productId: string): Promise<FaultSOP[]> {
  const response = await requestJson<FaultSOP[]>(`/api/knowledge/sops?product_id=${productId}`)

  if (!response.data) {
    throw new Error("服务未返回SOP列表。")
  }

  return response.data
}

export async function createSOP(payload: CreateSOPPayload): Promise<FaultSOP> {
  const response = await requestJson<FaultSOP>("/api/knowledge/sops", {
    method: "POST",
    body: JSON.stringify(payload),
  })

  if (!response.data) {
    throw new Error("创建SOP成功，但服务未返回SOP信息。")
  }

  return response.data
}

File: backend/api/routes/knowledge.py
from __future__ import annotations

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from backend.api.deps.database import get_db_session
from backend.api.schemas.knowledge import (
    ManualData,
    ManualUploadRequest,
    SOPCreateRequest,
    SOPData,
    SpecData,
    SpecImportRequest,
)
from backend.api.schemas.response import success_response

router = APIRouter(prefix="/knowledge", tags=["knowledge"])


@router.post("/manuals")
def upload_manual(
    payload: ManualUploadRequest,
    db_session: Session = Depends(get_db_session),
) -> dict:
    return success_response(
        message="手册上传成功",
        data=ManualData(
            id="placeholder",
            product_id=payload.product_id,
            title=payload.title,
            file_type=payload.file_type,
            status="pending",
            created_at="",
            updated_at="",
        ).model_dump(),
    )


@router.get("/manuals")
def list_manuals(
    product_id: str = Query(...),
    db_session: Session = Depends(get_db_session),
) -> dict:
    return success_response(message="获取手册列表成功", data=[])


@router.post("/specs")
def import_specs(
    payload: SpecImportRequest,
    db_session: Session = Depends(get_db_session),
) -> dict:
    return success_response(
        message="规格导入成功",
        data=[
            SpecData(
                id="placeholder",
                product_id=payload.product_id,
                spec_name=spec.spec_name,
                spec_value=spec.spec_value,
                spec_category=spec.spec_category,
            ).model_dump()
            for spec in payload.specs
        ],
    )


@router.get("/specs")
def list_specs(
    product_id: str = Query(...),
    db_session: Session = Depends(get_db_session),
) -> dict:
    return success_response(message="获取规格列表成功", data=[])


@router.post("/sops")
def create_sop(
    payload: SOPCreateRequest,
    db_session: Session = Depends(get_db_session),
) -> dict:
    return success_response(
        message="SOP创建成功",
        data=SOPData(
            id="placeholder",
            product_id=payload.product_id,
            fault_type=payload.fault_type,
            fault_category=payload.fault_category,
            symptoms=payload.symptoms,
            steps=payload.steps,
            resolution=payload.resolution,
        ).model_dump(),
    )


@router.get("/sops")
def list_sops(
    product_id: str = Query(...),
    db_session: Session = Depends(get_db_session),
) -> dict:
    return success_response(message="获取SOP列表成功", data=[])
